src/openms/source/ANALYSIS/ID/PrecursorPurity.cpp

```diff
@@ -144,64 +144,63 @@ namespace OpenMS
-    bool lastMS1(false);
-    int spectra_size = static_cast<int>(spectra.size());
-    bool break_loop(false);
+        // check for uniqueness of IDs by inserting initialized (0-value) scores into map
+        insert_return_value = purityscores.insert(std::pair<String, PrecursorPurity::PurityScores>(spectra[i].getNativeID(), PrecursorPurity::PurityScores()));
+        if (!insert_return_value.second)
+        {
+          OPENMS_LOG_WARN << "Warning: Input data not suitable for Precursor Purity computation. Duplicate Spectrum IDs. Precursor Purity info will not be calculated!\n";
+          return std::map<String, PrecursorPurity::PurityScores>();
+        }
+      }
+    }
 
 #pragma omp parallel for schedule(guided)
     for (int i = 0; i < spectra_size; ++i)
     {
-      // cannot return or break from OpenMP for loop, so continue to the end and return empty vector after loop
-      if (break_loop)
+      if (spectra[i].getMSLevel() == 2)
       {
-        continue;
-      }
-      // change current parent index if a new MS1 spectrum is reached
-      if (spectra[i].getMSLevel() == 1)
-      {
-        current_parent_index = i;
-      }
-      else if (spectra[i].getMSLevel() == 2)
-      {
-        // update next MS1 index, if it is lower than the current MS2 index
-        if (next_parent_index < i)
+        auto parent_spectrum_it = spectra.getPrecursorSpectrum(spectra.begin()+i);
+
+        // search for the next MS1 spectrum
+        int next_parent_index(0);
+        for (int j = i+1; j < spectra_size; ++j)
         {
-          for (int j = i+1; j < spectra_size; ++j)
+          if (spectra[j].getMSLevel() == 1)
           {
-            if (spectra[j].getMSLevel() == 1)
-            {
-              next_parent_index = j;
-              break;
-            }
-          }
-          // if the next MS1 index was not updated,
-          // the end of the PeakMap was reached right after this current group of MS2 spectra
-          if (next_parent_index < i)
-          {
-            lastMS1 = true;
+            next_parent_index = j;
+            break;
           }
         }
 
         PrecursorPurity::PurityScores score;
-        PrecursorPurity::PurityScores score1 = PrecursorPurity::computePrecursorPurity(spectra[current_parent_index], spectra[i].getPrecursors()[0], precursor_mass_tolerance, precursor_mass_tolerance_unit_ppm);
+        PrecursorPurity::PurityScores score1 = PrecursorPurity::computePrecursorPurity((*parent_spectrum_it), spectra[i].getPrecursors()[0], precursor_mass_tolerance, precursor_mass_tolerance_unit_ppm);
         // use default values of 0, if there is an MS2 spectrum without an MS1 after it (may happen for the last group of MS2 spectra)
         PrecursorPurity::PurityScores score2;
-        if (!lastMS1) // there is an MS1 after this MS2, combine values from previous and next MS1
+        if (next_parent_index > i) // there is an MS1 after this MS2, combine values from previous and next MS1
```

Comment:
why is this done?

Reply:
I got rid of the `lastMS1` variable as well, since it was also shared between threads.
If both flanking MS1 spectra are available, I combine the values from both. Otherwise I only use the parent MS1.

Reply:
I see. I would assume that this could go wrong sometimes (e.g., maybe polarity is changed or whatever). Would only use the parent

Reply:
@timosachsenberg If I remember correctly, this was originally your suggestion :)
I agree that it is not applicable in some cases. I will only use the parent then.

Reply:
hehe ok well then happy to review own suggestions

Reply:
@timosachsenberg  I also had to update some OpenPepXL test files this time, because most of the values PrecursorPurity returns now have changed. But no RNPxlSearch tests failed. Are there no MS1 spectra in you test data? That would be fine then. Otherwise this is unexpected.

Reply:
yes, they are plain MS2 spectra